File: app/src/main/java/nam/tran/architechture/di/module/ViewModelModule.java
package nam.tran.architechture.di.module;

import android.arch.lifecycle.ViewModel;
import android.arch.lifecycle.ViewModelProvider;

import dagger.Binds;
import dagger.Module;
import dagger.multibindings.IntoMap;
import nam.tran.architechture.view.main.activity.season.viewmodel.SoccerSeasonViewModel;
import nam.tran.architechture.view.main.activity.team.viewmodel.TeamViewModel;
import nam.tran.architechture.view.main.fragment.parent.season.viewmodel.SoccerSeasonChildViewModel;
import tran.nam.core.di.ViewModelFactory;
import tran.nam.core.di.inject.ViewModelKey;

@SuppressWarnings({"unused", "WeakerAccess"})
@Module
public abstract class ViewModelModule {

    @Binds
    abstract ViewModelProvider.Factory bindViewModelFactory(ViewModelFactory factory);

    @Binds
    @IntoMap
    @ViewModelKey(SoccerSeasonViewModel.class)
    abstract ViewModel bindSoccerSeasonViewModel(SoccerSeasonViewModel model);

    @Binds
    @IntoMap
    @ViewModelKey(TeamViewModel.class)
    abstract ViewModel bindTeamViewModel(TeamViewModel model);

    @Binds
    @IntoMap
    @ViewModelKey(SoccerSeasonChildViewModel.class)
    abstract ViewModel bindSoccerSeasonChildViewModel(SoccerSeasonChildViewModel model);
}

File: app/src/main/java/nam/tran/architechture/di/module/ViewModelModule.kt
package nam.tran.architechture.di.module

import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import dagger.Binds
import dagger.Module
import dagger.multibindings.IntoMap
import nam.tran.architechture.view.authen.viewmodel.LoginViewModel
import nam.tran.architechture.view.event.viewmodel.EventsViewModel
import nam.tran.architechture.view.repository.viewmodel.RepositoriesViewModel
import nam.tran.architechture.view.user.viewmodel.UserViewModel
import tran.nam.core.di.ViewModelFactory
import tran.nam.core.di.inject.ViewModelKey

@Module
abstract class ViewModelModule {

    @Binds
    internal abstract fun bindViewModelFactory(factory: ViewModelFactory): ViewModelProvider.Factory

    @Binds
    @IntoMap
    @ViewModelKey(LoginViewModel::class)
    internal abstract fun bindLoginViewModel(model: LoginViewModel): ViewModel

    @Binds
    @IntoMap
    @ViewModelKey(EventsViewModel::class)
    internal abstract fun bindEventsViewModel(model: EventsViewModel): ViewModel

    @Binds
    @IntoMap
    @ViewModelKey(RepositoriesViewModel::class)
    internal abstract fun bindRepositoriesViewModel(model: RepositoriesViewModel): ViewModel

    @Binds
    @IntoMap
    @ViewModelKey(UserViewModel::class)
    internal abstract fun bindUserViewModel(model: UserViewModel): ViewModel
}
